src/Codec/architecture.md

```diff
@@ -0,0 +1,112 @@
+
+The codec interface is comprised by two smaller interfaces: encoders and decoders. Both interfaces are marked as
+internal, as users are only expected to interact with the Codec interface. The interfaces are typed as generics through
+`@template` annotations, allowing for better type checking when they are used. Without type annotations, the interfaces
+are equivalent to the following:
+
+```php
+namespace MongoDB\Codec;
+
+interface Decoder
+{
+    public function canDecode(mixed $value): bool;
+
+    public function decode(mixed $value): mixed;
+
+    public function decodeIfSupported(mixed $value): mixed;
+}
+
+interface Encoder
+{
+    public function canEncode(mixed $value): bool;
+
+    public function encode(mixed $value): mixed;
+
+    public function encodeIfSupported(mixed $value): mixed;
+}
+```
+
+## Codec Interface
+
+The `Codec` interface combines decoding and encoding into a single interface. This will be used for most values except
+for documents where a more specific `DocumentCodec` is provided.
+
+The base interface supports encoding from a `NativeType` to a `BSONType` and back. Helper methods to determine whether a
+value is supported are provided. The `decodeIfSupported` and `encodeIfSupported` methods are useful to have a codec
+encode or decode a value only if it is supported. If it is not supported, the original value is returned.
+
+```php
+namespace MongoDB\Codec;
+
+interface Codec extends Decoder, Encoder
+{
+}
+```
+
+## Document Codec
+
+The document codec is special as it is guaranteed to always encode to a BSON document instance and decode to a PHP
+object. Document codecs will be used by `MongoDB\Collection` instances to automatically decode BSON data into PHP
+objects when reading data, and to encode PHP objects when inserting or replacing data.
+
+```php
+namespace MongoDB\Codec;
+
+use MongoDB\BSON\Document;
+
+/** 
+ * @template ObjectType of object
+ * @extends Codec<ObjectType, Document> 
+ */
+interface DocumentCodec extends Codec
+{
+}
+```
+
+## Built-in codecs
+
+By default, two codecs are provided: an `ArrayCodec` and an `ObjectCodec`. These two codecs are used to recursively
+encode and decode values in arrays and `stdClass` instances, respectively. When encoding or decoding an object,
+`ObjectCodec` only handles public properties of the object and ignores private and protected properties.
+
+## Future Work
+
+### Using Codecs
+
+The `MongoDB\Collection` class and all operations that work with documents now take a `codec` option. This option is
+passed along to the various operations that already take a `typeMap` option. Collections only support a `DocumentCodec`
+instance to guarantee that data always encodes to a BSON document and decodes to a PHP object.
+
+All operations that return documents will use the codec to decode the documents into PHP objects. This includes
+the various `find` and `findAndModify` operations in collections as well as the `aggregate` and `watch` operations in
```

Comment:
Ah, yes. `watch` definitely needs a different codec, `aggregate` is less straightforward. `distinct` is also a special case that deserves some thought.

Given the vastly different use cases, I believe the best course of action would be to not inherit the codec for those operations:
* `aggregate` is expected to produce significantly different results, so it should not inherit the default document codec
* `distinct` returns a list of values from the document, so we know that the default document codec is not the correct choice
* While `watch` has a consistent return document, this can be changed further using the `pipeline` option for `watch`. With that in mind, the same restrictions apply as for `aggregate` and users should specify their own codec.

For watch, we could provide a default codec and result models that expose rich versions of change stream events. However, I would track that as future scope.

Reply:
> `distinct` returns a list of values from the document, so we know that the default document codec is not the correct choice

Thinking about this some more, the codec implementation here doesn't care about field paths at all. In that case, wouldn't it still be compatible with `distinct` if the command is returning values directly from the document? Presumably, the same strategy would apply for whatever types are returned by `distinct` as would apply to the root document itself, no?

Apologies if this was something we already discussed in the previous PR, in which case I've forgotten.

Reply:
@alcaeus: outstanding question above. It doesn't affect my review, but I'd be curious to know before this gets picked up in a future PR.

Reply:
Sorry for missing this one.

Outside of a codec, field paths are completely irrelevant. For `distinct`, this means that it should not inherit the codec from the collection by default as it will not return the usual result document. Users should instead pass the appropriate codec they need when reading values for `distinct`.

PHPLIB-1182 will be adding codec options to all operations classes. I'll make sure to add appropriate examples for `distinct` so we can discuss this then.